tower-http/src/timeout/body.rs

```diff
@@ -0,0 +1,373 @@
+use pin_project_lite::pin_project;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::{sleep, Sleep};
+use tower_layer::Layer;
+use tower_service::Service;
+
+pin_project! {
+    /// Wrapper around a [`http_body::Body`] to time out if data is not ready within the specified duration.
+    pub struct TimeoutBody<B> {
+        timeout: Duration,
+        #[pin]
+        sleep: Option<Sleep>,
+        #[pin]
+        body: B,
+    }
+}
+
+impl<B> TimeoutBody<B> {
+    /// Creates a new [`TimeoutBody`].
+    pub fn new(timeout: Duration, body: B) -> Self {
+        TimeoutBody {
+            timeout,
+            sleep: None,
+            body,
+        }
+    }
+}
+
+impl<B> Body for TimeoutBody<B>
+where
+    B: Body,
+{
+    type Data = B::Data;
+    type Error = Box<dyn std::error::Error + Send + Sync>;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        let mut this = self.project();
+
+        // Start the `Sleep` if not active.
+        let sleep_pinned = if let Some(some) = this.sleep.as_mut().as_pin_mut() {
+            some
+        } else {
+            this.sleep.set(Some(sleep(*this.timeout)));
+            this.sleep.as_mut().as_pin_mut().unwrap()
+        };
+
+        // Error if the timeout has expired.
+        match sleep_pinned.poll(cx) {
+            Poll::Pending => (),
+            Poll::Ready(()) => return Poll::Ready(Some(Err(Box::new(TimeoutError)))),
+        }
+
+        // Check for body data.
+        match this.body.poll_data(cx) {
+            Poll::Ready(data) => {
+                // Some data is ready. Reset the `Sleep`...
+                this.sleep.set(Some(sleep(*this.timeout)));
+
+                // ...then `poll` it to get awoken.
+                if let Poll::Ready(_) = this.sleep.as_pin_mut().unwrap().poll(cx) {
+                    return Poll::Ready(Some(Err(Box::new(TimeoutError))));
+                }
+                Poll::Ready(
+                    data.transpose()
+                        .map_err(|_| Box::new(TimeoutError).into())
+                        .transpose(),
+                )
+            }
+            Poll::Pending => Poll::Pending,
+        }
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<http::HeaderMap>, Self::Error>> {
+        let this = self.project();
+
+        // Error if the timeout has expired.
+        match this
+            .sleep
+            .as_pin_mut()
+            .expect("poll_data was not called")
+            .poll(cx)
+        {
+            Poll::Pending => (),
+            Poll::Ready(()) => return Poll::Ready(Err(Box::new(TimeoutError))),
+        }
+
+        this.body
+            .poll_trailers(cx)
+            .map_err(|_| Box::new(TimeoutError).into())
+    }
+}
+
+/// Error for [`TimeoutBody`].
+#[derive(Debug)]
+struct TimeoutError;
```

Comment:
This struct should be pub with a `(())` type inside so it can't be constructed outside the crate. We should look at how other middleware in this crate handle errors.